### ISSUE TITLE
levelByDirection inconsistent

### ISSUE THREAD

**Issue by heinrichmartin** [almende/vis#1601]
Hi again,
this time, my observation is rather vague. I noticed vertical nodes in my graph (not a tree) with hierarchical layout and sortMethod directed.
It looks like the implementation visits each edge only once - which might give inconsistent results depending on the sequence when the graph is not a tree.

I implemented a version that does not allow vertical edges.
* Note that neighbors might be more than one level apart from each other in such graphs.
* My implementation will probably fail on cycles (endless recursion?).
* I did not dig into your release procedure, but hacked dist/vis.js directly.
* It is probably not possible to implement my algorithm with the unreleased custom sortMethod (unless the callback can report that nodes must be revisited).

I understand that you might just close this as "won't fix", because I did not follow your guidlines for contributing. But it is now available for refinement at least - and the need is documented :-)

```patch
Index: dist/vis.js
===================================================================
--- dist/vis.js
+++ dist/vis.js
@@ -41132,16 +41132,28 @@

         var minLevel = 10000;
         var levelByDirection = function levelByDirection(nodeA, nodeB, edge) {
-          var levelA = _this5.hierarchicalLevels[nodeA.id];
-          // set initial level
-          if (levelA === undefined) {
-            _this5.hierarchicalLevels[nodeA.id] = minLevel;
-          }
+          var from;
+          var to;
           if (edge.toId == nodeB.id) {
-            _this5.hierarchicalLevels[nodeB.id] = _this5.hierarchicalLevels[nodeA.id] + 1;
+            from = nodeA.id;
+            to   = nodeB.id;
           } else {
-            _this5.hierarchicalLevels[nodeB.id] = _this5.hierarchicalLevels[nodeA.id] - 1;
+            from = nodeB.id;
+            to   = nodeA.id;
           }
+          var levelFrom = _this5.hierarchicalLevels[from];
+          var levelTo   = _this5.hierarchicalLevels[to];
+          if (levelFrom === undefined && levelTo === undefined) {
+            _this5.hierarchicalLevels[from] = minLevel;
+            _this5.hierarchicalLevels[to]   = minLevel  + 1;
+          } else if (levelFrom === undefined) {
+            _this5.hierarchicalLevels[from] = levelTo   - 1;
+          } else if (levelTo === undefined) {
+            _this5.hierarchicalLevels[to]   = levelFrom + 1;
+          } else if (levelFrom >= levelTo) {
+            _this5.hierarchicalLevels[to]   = levelFrom + 1;
+            _this5._crawlNetwork(levelByDirection, to);
+          }
         };
         this._crawlNetwork(levelByDirection);
         this._setMinLevelToZero();
```

**Comment by AlexDM0** [comment]
> My implementation will probably fail on cycles (endless recursion?).

this will happen. We cannot allow this to be in the code, you'll either have to implement it correctly and do a pull or argue what you expect to happen. The results are consistent since there is no random factor in there. If you can give a dataset which shows your idea we have something to develop against.

Regards
